lib/upnp/upnp.go

```diff
@@ -141,24 +171,44 @@ func Discover(ctx context.Context, _, timeout time.Duration) []nat.Device {
 
 // Search for UPnP InternetGatewayDevices for <timeout> seconds.
 // The order in which the devices appear in the result list is not deterministic
-func discover(ctx context.Context, intf *net.Interface, deviceType string, timeout time.Duration, results chan<- nat.Device) {
-	ssdp := &net.UDPAddr{IP: []byte{239, 255, 255, 250}, Port: 1900}
+func discover(ctx context.Context, intf *net.Interface, deviceType string, timeout time.Duration, results chan<- nat.Device, ip6 bool) {
+	var ssdp net.UDPAddr
+	var template string
+	if ip6 {
+		ssdp = net.UDPAddr{IP: []byte{0xFF, 0x05, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0C}, Port: 1900}
+
+		template = `M-SEARCH * HTTP/1.1
+HOST: [FF05::C]:1900
+ST: %s
+MAN: "ssdp:discover"
+MX: %d
+USER-AGENT: syncthing/%s
 
-	tpl := `M-SEARCH * HTTP/1.1
+`
+	} else {
+		ssdp = net.UDPAddr{IP: []byte{239, 255, 255, 250}, Port: 1900}
+
+		template = `M-SEARCH * HTTP/1.1
 HOST: 239.255.255.250:1900
 ST: %s
 MAN: "ssdp:discover"
 MX: %d
-USER-AGENT: syncthing/1.0
+USER-AGENT: syncthing/%s
 
 `
-	searchStr := fmt.Sprintf(tpl, deviceType, timeout/time.Second)
+	}
+
+	searchStr := fmt.Sprintf(template, deviceType, timeout/time.Second, build.Version)
 
 	search := []byte(strings.ReplaceAll(searchStr, "\n", "\r\n") + "\r\n")
 
 	l.Debugln("Starting discovery of device type", deviceType, "on", intf.Name)
 
-	socket, err := net.ListenMulticastUDP("udp4", intf, &net.UDPAddr{IP: ssdp.IP})
+	proto := "udp4"
+	if ip6 {
+		proto = "udp6"
+	}
+	socket, err := net.ListenMulticastUDP(proto, intf, &net.UDPAddr{IP: ssdp.IP})
 	if err != nil {
 		l.Debugln("UPnP discovery: listening to udp multicast:", err)
```

Comment:
This fails for me on Windows:

```
UPnP discovery: listening to udp multicast: listen udp6 [ff05::c]:0: setsockopt: not supported by windows
```

Reply:
Maybe limited to the interface?

```
[NTMID] 2023/10/02 17:30:23.963447 upnp.go:205: DEBUG: Starting discovery of device type urn:schemas-upnp-org:device:InternetGatewayDevice:2 on WLAN
[NTMID] 2023/10/02 17:30:23.964446 upnp.go:205: DEBUG: Starting discovery of device type urn:schemas-upnp-org:device:InternetGatewayDevice:2 on Loopback Pseudo-Interface 1
[NTMID] 2023/10/02 17:30:23.964446 upnp.go:213: DEBUG: UPnP discovery: listening to udp multicast: listen udp6 [ff05::c]:0: setsockopt: not supported by windows
[NTMID] 2023/10/02 17:30:23.964446 upnp.go:205: DEBUG: Starting discovery of device type urn:schemas-upnp-org:device:InternetGatewayDevice:2 on WLAN
[NTMID] 2023/10/02 17:30:23.982272 upnp.go:218: DEBUG: Sending search request for device type urn:schemas-upnp-org:device:InternetGatewayDevice:2 on WLAN
[NTMID] 2023/10/02 17:30:23.983272 upnp.go:218: DEBUG: Sending search request for device type urn:schemas-upnp-org:device:InternetGatewayDevice:2 on Loopback Pseudo-Interface 1
[NTMID] 2023/10/02 17:30:23.983272 upnp.go:228: DEBUG: Listening for UPnP response for device type urn:schemas-upnp-org:device:InternetGatewayDevice:2 on Loopback Pseudo-Interface 1
[NTMID] 2023/10/02 17:30:23.983272 upnp.go:228: DEBUG: Listening for UPnP response for device type urn:schemas-upnp-org:device:InternetGatewayDevice:2 on WLAN
```

Reply:
Added a few debug statements to rule out any special interface funzies. For my primary wifi interface this fails outright.

Should we switch to `x/net/ipv{4,6}` instead?

Reply:
I'm currently looking at it, although I probably won't have much time in the next few days. My understanding so far is that the Go standard library does three syscalls [here](https://cs.opensource.google/go/go/+/master:src/net/udpsock_posix.go;l=274;drc=bc15070085ec417d4254f8a4eda62b42de88fb37) and one of them might fail on Windows, although I don't know which one or why. x/net seems to have a different implementation so it may be worth a try, although it's a more involved API for doing the same thing as `ListenMulticastUDP` (Join the multicast group, disable loopback so we don't get our own message as an UPnP advertisement, broadcast the solicitation, receive the answer)

[This example](https://github.com/golang/net/blob/88194ad8ab44a02ea952c169883c3f57db6cf9f4/ipv6/example_test.go#L48C21-L48C21) might be useful for that.

Reply:
Is this a regression? If not, can we exclude it from the scope of this PR please. I haven't looked at difference of net vs x/net/ipv4, but if it's more complex, there's even the question if we want that complexity. Which regardless of the answer, will delay this PR again. Also a/the solution might then also be to file an issue for net.

Reply:
The IPv4 multicast still works. I haven't tested the current state of the PR on Linux, but at least for Windows the IPv6 multicast listener fails.

Reply:
Can't say whether it's a regression, but I can do some more testing this weekend at the latest, on Linux. Would be great to know what look for regarding this question.

Reply:
Pretty easy to spot. The SSDP search simply fails and syncthing is unable to find suitable IPv6 enabled routers.

Reply:
OK so I think it's actually discovering my router over IPv6 and trying to open ports as well.

All in all, it's still very flaky, probably because of our interaction with the router firmware's bugs.  I had one Wi-Fi and one Ethernet connection to the same router, both with valid IPv4 and IPv6 addresses. TCP and QUIC listeners configured for 0.0.0.0:22123 and another separate listener on tcp4://0.0.0.0:22124.  For a short while, I had one successful IPv6 mapping on the wlan0 interface and two IPv4 mappings on Ethernet, as reported by my router.  After a few seconds, the Ethernet mappings were both gone again, but IPv6 on Wi-Fi stayed active.  Must be some later action we try that somehow overrides the previously successful mappings.

I don't know how to proceed with this really, how much of that is a regression and what exactly had worked before.

Reply:
I don't think there's any code that actively removes a mapping or a pinhole, we just let them expire. With all the buggy router firmware out there, I'm not really convinced that our implementation is to blame in this case. UPnP should be treated as a best effort. After creating a mapping/pinhole, we are at the mercy of the manufacturers' UPnP implementations and must hope that there are no unexpected surprises.

Given that the mapping disappeared only a few seconds after it was created, we can at least rule out a faulty update routine on our end.

Pro tip: run Wireshark while you're testing the PR. You can use the `ssdp` filter to only display discovery traffic and `http` to display the UPnP requests.

Reply:
I haven't looked deeply into it. Just from the timing, and looking at the router GUI while the process was not yet finished, I think that our numerous attempts at doing everything over every possible protocol might be to blame. If we send a request that (from the router's possibly buggy interpretation) conflicts with a previous entry, we get an error response and the previously successful mapping is removed as a side effect.

Will need to find some time to really understand the sequence of events in the logs. Or I can send them to you if you're interested.